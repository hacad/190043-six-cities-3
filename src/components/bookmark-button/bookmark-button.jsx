import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {Operation} from "../../reducers/data/reducer.js";
import history from "../../history.js";
import ErrorLabel from "../error-label/error-label.jsx";

function BookmarkButton(props) {
  const {offerId, isFavorite, toggleFavorite, errors,
    iconWidth, iconHeight, classNamePrefix,
    isAuthorized} = props;
  const className = `${classNamePrefix}__bookmark-button button ${isFavorite && isAuthorized ? `${classNamePrefix}__bookmark-button--active` : ``}`;

  const renderToggleFavoriteError = errors && errors[`toggleFavorite`]
    ? <ErrorLabel txtLbl={errors[`toggleFavorite`]} htmlFor="toggleFavorite" />
    : ``;
  return (
    <button
      className={className}
      id="toggleFavorite"
      type="button"
      onClick={() => isAuthorized ? toggleFavorite(offerId, !isFavorite) : history.push(`/login`)}>
      <svg className="place-card__bookmark-icon property__bookmark-icon" width={iconWidth} height={iconHeight}>
        <use xlinkHref="#icon-bookmark"></use>
      </svg>
      <span className="visually-hidden">{isFavorite ? `In bookmarks` : `To bookmarks`}</span>
      {renderToggleFavoriteError}
    </button>
  );
}

BookmarkButton.propTypes = {
  offerId: PropTypes.number.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  toggleFavorite: PropTypes.func.isRequired,
  iconWidth: PropTypes.number.isRequired,
  iconHeight: PropTypes.number.isRequired,
  classNamePrefix: PropTypes.string.isRequired,
  isAuthorized: PropTypes.bool.isRequired,
  errors: PropTypes.shape({
    toggleFavorite: PropTypes.string,
  })
};

function mapDispatchToProps(dispatch, ownProps) {
  return {
    toggleFavorite: (offerId, isFavorite) => {
      dispatch(Operation.toggleFavorite(offerId, isFavorite))
        .catch((error) => {
          ownProps.errors = {"toggleFavorite": error.response && error.response.data
            ? error.response.data.error
            : error};
        });
    }
  };
}

export {BookmarkButton};

export default connect(null, mapDispatchToProps)(BookmarkButton);
