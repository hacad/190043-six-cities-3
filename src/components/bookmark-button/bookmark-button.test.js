import React from "react";
import renderer from "react-test-renderer";
import {BookmarkButton} from "./bookmark-button.jsx";

it(`BookmarkButton correctly renders after relaunch`, () => {
  const bookmarkButton = renderer.create(
      <BookmarkButton
        offerId={1}
        isFavorite={true}
        toggleFavorite={jest.fn()}
        iconWidth={18}
        iconHeight={19}
        classNamePrefix="place-card"
        isAuthorized={true}
      />).toJSON();

  expect(bookmarkButton).toMatchSnapshot();
});
