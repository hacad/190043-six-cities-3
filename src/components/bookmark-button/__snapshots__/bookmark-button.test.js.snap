// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkButton correctly renders after relaunch 1`] = `
<button
  className="place-card__bookmark-button button place-card__bookmark-button--active"
  id="toggleFavorite"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon property__bookmark-icon"
    height={19}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    In bookmarks
  </span>
  
</button>
`;
